File: src-tauri/src/types/messages.rs
/// ParsedMessage — the application's internal representation after parsing raw JSONL entries.

use serde::{Deserialize, Serialize};
use serde_json::Value;

use super::jsonl::ContentBlock;

// =============================================================================
// Tool Types
// =============================================================================

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ToolCall {
    pub id: String,
    pub name: String,
    pub input: Value,
    pub is_task: bool,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub task_description: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub task_subagent_type: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ToolResult {
    pub tool_use_id: String,
    pub content: Value,
    pub is_error: bool,
}

// =============================================================================
// Message Content (string | ContentBlock[])
// =============================================================================

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(untagged)]
pub enum ParsedMessageContent {
    Text(String),
    Blocks(Vec<ContentBlock>),
}

// =============================================================================
// Parsed Message
// =============================================================================

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ParsedMessage {
    pub uuid: String,
    pub parent_uuid: Option<String>,
    #[serde(rename = "type")]
    pub message_type: String,
    /// ISO-8601 timestamp string (frontend handles conversion)
    pub timestamp: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub role: Option<String>,
    pub content: ParsedMessageContent,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub usage: Option<TokenUsage>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub model: Option<String>,
    // Metadata
    #[serde(skip_serializing_if = "Option::is_none")]
    pub cwd: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub git_branch: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub agent_id: Option<String>,
    pub is_sidechain: bool,
    pub is_meta: bool,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub user_type: Option<String>,
    // Tool information
    pub tool_calls: Vec<ToolCall>,
    pub tool_results: Vec<ToolResult>,
    #[serde(skip_serializing_if = "Option::is_none")]
    #[serde(rename = "sourceToolUseID")]
    pub source_tool_use_id: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    #[serde(rename = "sourceToolAssistantUUID")]
    pub source_tool_assistant_uuid: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub tool_use_result: Option<Value>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub is_compact_summary: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub request_id: Option<String>,
}

// =============================================================================
// Token Usage
// =============================================================================

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TokenUsage {
    #[serde(default)]
    pub input_tokens: u64,
    #[serde(default)]
    pub output_tokens: u64,
    #[serde(default)]
    pub cache_read_input_tokens: Option<u64>,
    #[serde(default)]
    pub cache_creation_input_tokens: Option<u64>,
}

// =============================================================================
// Message Category
// =============================================================================

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub enum MessageCategory {
    User,
    System,
    HardNoise,
    Ai,
    Compact,
}

File: internal/domain/messages.go
package domain

import "encoding/json"

// ToolCall — an extracted tool_use block.
type ToolCall struct {
	ID               string   `json:"id"`
	Name             string   `json:"name"`
	Input            RawValue `json:"input"`
	IsTask           bool     `json:"isTask"`
	TaskDescription  *string  `json:"taskDescription,omitempty"`
	TaskSubagentType *string  `json:"taskSubagentType,omitempty"`
}

// ToolResult — an extracted tool_result block.
type ToolResult struct {
	ToolUseID string   `json:"toolUseId"`
	Content   RawValue `json:"content"`
	IsError   bool     `json:"isError"`
}

// ParsedMessageContent mirrors the Rust untagged enum
// `Text(String) | Blocks(Vec<ContentBlock>)`.
type ParsedMessageContent struct {
	Text   *string
	Blocks []ContentBlock
}

func (c ParsedMessageContent) MarshalJSON() ([]byte, error) {
	if c.Text != nil {
		return json.Marshal(*c.Text)
	}
	return json.Marshal(c.Blocks)
}

func (c *ParsedMessageContent) UnmarshalJSON(b []byte) error {
	var s string
	if err := json.Unmarshal(b, &s); err == nil {
		c.Text = &s
		return nil
	}
	return json.Unmarshal(b, &c.Blocks)
}

// ParsedMessage — the internal representation after parsing a JSONL entry.
// parent_uuid has NO skip_serializing_if in Rust → it serializes as null when
// absent (so *string WITHOUT omitempty). Tool slices are always arrays.
type ParsedMessage struct {
	UUID        string               `json:"uuid"`
	ParentUUID  *string              `json:"parentUuid"`
	MessageType string               `json:"type"`
	Timestamp   string               `json:"timestamp"`
	Role        *string              `json:"role,omitempty"`
	Content     ParsedMessageContent `json:"content"`
	Usage       *TokenUsage          `json:"usage,omitempty"`
	Model       *string              `json:"model,omitempty"`

	Cwd       *string `json:"cwd,omitempty"`
	GitBranch *string `json:"gitBranch,omitempty"`
	AgentID   *string `json:"agentId,omitempty"`
	IsSidechain bool  `json:"isSidechain"`
	IsMeta      bool  `json:"isMeta"`
	UserType  *string `json:"userType,omitempty"`

	ToolCalls   []ToolCall `json:"toolCalls"`
	ToolResults []ToolResult `json:"toolResults"`

	SourceToolUseID         *string `json:"sourceToolUseID,omitempty"`
	SourceToolAssistantUUID *string `json:"sourceToolAssistantUUID,omitempty"`
	ToolUseResult           RawValue `json:"toolUseResult,omitempty"`
	IsCompactSummary        *bool   `json:"isCompactSummary,omitempty"`
	RequestID               *string `json:"requestId,omitempty"`
	Subtype                 *string `json:"subtype,omitempty"`
	EventData               *SystemEventData `json:"eventData,omitempty"`
}

// TokenUsage has NO rename_all in Rust → fields keep snake_case names. The
// cache_* fields have no skip_serializing_if, so they emit null when absent.
type TokenUsage struct {
	InputTokens             uint64  `json:"input_tokens"`
	OutputTokens            uint64  `json:"output_tokens"`
	CacheReadInputTokens    *uint64 `json:"cache_read_input_tokens"`
	CacheCreationInputTokens *uint64 `json:"cache_creation_input_tokens"`
}

// SystemEventData — typed payload for system/event messages.
type SystemEventData struct {
	Subtype string `json:"subtype"`

	ErrorStatus  *uint16  `json:"errorStatus,omitempty"`
	ErrorType    *string  `json:"errorType,omitempty"`
	ErrorMessage *string  `json:"errorMessage,omitempty"`
	RetryAttempt *uint32  `json:"retryAttempt,omitempty"`
	MaxRetries   *uint32  `json:"maxRetries,omitempty"`
	RetryInMs    *float64 `json:"retryInMs,omitempty"`

	BridgeContent *string `json:"bridgeContent,omitempty"`
	BridgeURL     *string `json:"bridgeUrl,omitempty"`

	WrittenPaths *[]string `json:"writtenPaths,omitempty"`
	MemoryVerb   *string   `json:"memoryVerb,omitempty"`

	DurationMs *float64 `json:"durationMs,omitempty"`

	Operation     *string `json:"operation,omitempty"`
	QueuedContent *string `json:"queuedContent,omitempty"`
}

// MessageCategory — classification result (serde camelCase). Not part of the
// SessionDetail gate output; used by the W3 classifier.
type MessageCategory string

const (
	CategoryUser      MessageCategory = "user"
	CategorySystem    MessageCategory = "system"
	CategoryHardNoise MessageCategory = "hardNoise"
	CategoryAi        MessageCategory = "ai"
	CategoryCompact   MessageCategory = "compact"
	CategoryEvent     MessageCategory = "event"
)
